diff --git a/src/sdl.c b/src/sdl.c
--- a/src/sdl.c
+++ b/src/sdl.c
@@ -73,6 +73,7 @@
   SDL_Event event;
 
   SDL_SetRelativeMouseMode(SDL_TRUE);
+  SDL_SetWindowKeyboardGrab(window, SDL_TRUE);
 
   while(!done && SDL_WaitEvent(&event)) {
     switch (sdlinput_handle_event(window, &event)) {
@@ -84,12 +85,14 @@
       SDL_SetWindowFullscreen(window, fullscreen_flags);
       break;
     case SDL_MOUSE_GRAB:
-      SDL_ShowCursor(SDL_ENABLE);
+      SDL_ShowCursor(SDL_DISABLE);
       SDL_SetRelativeMouseMode(SDL_TRUE);
+      SDL_SetWindowKeyboardGrab(window, SDL_TRUE);
       break;
     case SDL_MOUSE_UNGRAB:
       SDL_SetRelativeMouseMode(SDL_FALSE);
-      SDL_ShowCursor(SDL_DISABLE);
+      SDL_ShowCursor(SDL_ENABLE);
+      SDL_SetWindowKeyboardGrab(window, SDL_FALSE);
       break;
     default:
       if (event.type == SDL_QUIT)
