diff -ruN a/src/alejandra/src/rules/bin_op.rs b/src/alejandra/src/rules/bin_op.rs
--- a/src/alejandra/src/rules/bin_op.rs
+++ b/src/alejandra/src/rules/bin_op.rs
@@ -11,14 +11,16 @@
     let second = children.next().unwrap();
     let third = children.next().unwrap();
 
-    let vertical = build_ctx.vertical
-        || first.has_inline_comment
+    let has_trivia = first.has_inline_comment
         || first.has_trivialities
         || second.has_inline_comment
         || second.has_trivialities
         || third.has_inline_comment
         || third.has_trivialities;
 
+    let merge_operator = second.element.to_string() == "//";
+    let vertical = has_trivia || (build_ctx.vertical && !merge_operator);
+
     // first
     if vertical {
         let kind = first.element.kind();
@@ -84,7 +86,7 @@
     }
 
     // third
-    if vertical {
+    if vertical || merge_operator {
         steps.push_back(crate::builder::Step::FormatWider(third.element));
     } else {
         steps.push_back(crate::builder::Step::Format(third.element));
diff -ruN a/src/alejandra/src/rules/key_value.rs b/src/alejandra/src/rules/key_value.rs
--- a/src/alejandra/src/rules/key_value.rs
+++ b/src/alejandra/src/rules/key_value.rs
@@ -93,6 +93,10 @@
         } else if matches!(
             child_expr.kind(),
             rnix::SyntaxKind::NODE_ASSERT
+                | rnix::SyntaxKind::NODE_APPLY
+                | rnix::SyntaxKind::NODE_BIN_OP
+                | rnix::SyntaxKind::NODE_IDENT
+                | rnix::SyntaxKind::NODE_SELECT
                 | rnix::SyntaxKind::NODE_ATTR_SET
                 | rnix::SyntaxKind::NODE_PAREN
                 | rnix::SyntaxKind::NODE_LAMBDA
diff -ruN a/src/alejandra/src/rules/let_in.rs b/src/alejandra/src/rules/let_in.rs
--- a/src/alejandra/src/rules/let_in.rs
+++ b/src/alejandra/src/rules/let_in.rs
@@ -18,7 +18,7 @@
     let vertical = items_count > 1
         || children.has_comments()
         || children.has_newlines()
-        || build_ctx.vertical;
+        || (build_ctx.vertical && items_count > 1);
 
     // let
     let child = children.get_next().unwrap();
diff -ruN a/src/alejandra/src/rules/pattern.rs b/src/alejandra/src/rules/pattern.rs
--- a/src/alejandra/src/rules/pattern.rs
+++ b/src/alejandra/src/rules/pattern.rs
@@ -27,7 +27,7 @@
 
     let arguments_count = pattern.arguments.len();
 
-    let arguments_count_for_tall = if has_ellipsis { 2 } else { 1 };
+    let arguments_count_for_tall = if has_ellipsis { 4 } else { 3 };
 
     let vertical = has_comments
         || arguments_count > arguments_count_for_tall
