diff -ruN a/src/alejandra/src/rules/pattern.rs b/src/alejandra/src/rules/pattern.rs
--- a/src/alejandra/src/rules/pattern.rs
+++ b/src/alejandra/src/rules/pattern.rs
@@ -27,7 +27,7 @@
 
     let arguments_count = pattern.arguments.len();
 
-    let arguments_count_for_tall = if has_ellipsis { 2 } else { 1 };
+    let arguments_count_for_tall = if has_ellipsis { 4 } else { 3 };
 
     let vertical = has_comments
         || arguments_count > arguments_count_for_tall
